filament/backend/src/opengl/OpenGLContext.cpp

```diff
@@ -521,4 +521,21 @@ void OpenGLContext::deleteVertexArrays(GLsizei n, const GLuint* arrays) noexcept
     }
 }
 
+void OpenGLContext::resetStateTracking() noexcept {
+    // Store off strings instead of re-calling glGetString()
+    char const* vendor = state.vendor;
+    char const* renderer = state.renderer;
+    char const* version = state.version;
+    char const* shader = state.shader;
+
+    // Reset state to defaults
+    state = State {};
```

Comment:
I don't understand why this works. Nothing guarantees that the default state matches the current GL state left by the app.

I think, unfortunately, what you need to do is go through every bit of state in the `State` structure and forcibly set the corresponding GL state to that.

Reply:
Good point. This worked for the bits I had good repros for - bound textures, viewport, fbo, since they got set to null values & re-set. I should've realized that won't work for most of the rest of the state. Will fix.

Reply:
I think I've taken care of this, but I may have misunderstood some piece - `state.vao` and `state.textures` in particular I'm not completely confident about.

Reply:
I *think* technically, you don't need to reset the  `State` structure anymore, since you're re-syncing the gl state to it.

Reply:
I think there are three exceptions right now:
1. I'm calling `bindVertexArray(nullptr);` - this looks fairly straightforward to change:
```cpp
glBindVertexArray(state.vao.p->vao);
if (UTILS_UNLIKELY(bugs.vao_doesnt_store_element_array_buffer_binding)) {
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, state.vao.p->elementArray);
}
```

2. Resetting texture units is just binding 0 everywhere. If I'm understanding correctly, I'll need to lookup the unit for the call to `glBindSampler`: `state.textures.units[unit].sampler`? And then for `glBindTexture()`: ```
```cpp
int targets_index = getIndexForTextureTarget(target);
glBindTexture(target, state.textures.units[unit].targets[targets_index].texture_id`);
```

3. I broke out setting the initial `glDisable()/glEnable()`, `glHint` and `glClipControl` into `OpenGLContext::setDefaultState()`. I can move those back into the constructor & manually run through `state.enables`.  I assume I'll need to run through the values in `getIndexForCap()` and set those based on what's set in `state.enables.caps`?

I'd also then change `Driver::resetStateTracking()` to `Driver::resetState()`.

If those proposed changes look good I'll update the PR.

Reply:
@pixelflinger I'm attempting to make these changes - however, I'm running into issues with the texture state that I think is part of a different issue.

`OpenGLDriver` uses  `OpenGLContext::MAX_TEXTURE_UNIT_COUNT - 1` in a bunch of places when binding textures to upload them. However, that constant is larger than the allowed number of units available in WebGL, at least on the browsers I've tried on Windows. This causes warnings to be printed (Firefox): `WebGL warning: activeTexture: TEXTURE61 must be < MAX_COMBINED_TEXTURE_IMAGE_UNITS (32).`. I think this also means the state tracking ends up in an invalid state, because the active texture unit is failing to be set. I'll open a ticket about this.

Reply:
yes, we fixed that one recently. it was my bad.

Reply:
Ah, I should've synced. Thanks